### DISCUSSION TITLE
How to add temporally variable ozone and CO2 input in single point simulation?

### DISCUSSION THREAD

**Discussion by jyoti1singh** [ESCOMP/CTSM#1848] [Q&A] [answered]
Hi all,

I am trying to do a single-point simulation in CLM5 for temporally variable ozone input data. Earlier only a constant value of ozone concentration was taken as input, but because of [this new development](https://github.com/ESCOMP/CTSM/pull/1837) now it is possible to prescribe spatially and temporally variable ozone concentration. I am working on making CLM5 a part of the Agricultural Model Intercomparison and Improvement Project (AgMIP) ozone initiative. I have hourly data on ozone concentration for the entire season for multiple experimental sites that I have to use for calibration and validation to assess the ozone impact on winter and spring wheat.

I do not know how to prepare ozone as a temporally varying input. Is it similar to adding CO2 concentration? Please help me set up the single-point simulation taking O3 and CO2 as a temporally variable input.

**Comment by adrifoster** [comment]
@swensosc, @ekluzek, @wwieder Do we have examples for how to do this for other DATM streams? I am thinking something like the NEON tower sites would be relevant?

**Comment by ekluzek** [comment, marked as answer]
@jyoti1singh what you need to do is to make sure both CO2 and O3 is turned on from datm to CLM. And then you swap out the global datasets pointed to in the datm.streams.xml file under CaseDocs, with the new files you want to point to using user_nl_datm_streams. CO2 is normally turned on by datm for Hist compsets.

So the xml variables:

DATM_CO2_TSERIES
DATM_PRESO3

Both need to be set to something. And then you can modify their streams with the user_nl_datm_streams file. So you'll probably need to set the filenames with something like...

```
preso3.hist:datafiles = <filename>
co2tseries.20tr:datafiles = <filename>
```
You might also need to change the mesh file to go with the grid of your data, as well as the start and end years for it. You don't need to specify anything about the temporal frequency of the data, it'll just figure out what's on the dataset and use it that way.

**Comment by jyoti1singh** [comment]
@ekluzek I have done as you said in datm.streams.xml for Ozone and also added a user_datm.streams.txt.OZONE. By default, the ozone file is of monthly temporal resolution. However, when I built the case. preso3.SSP2-4.5 again went back to its default monthly file in datm.streams.xml. probably;y because I did not update the user_nl_datm_streams file. Is user_nl_datm_streams a new addition because I never used it before? This is my case folder /glade/work/jyotis/O3_CLM5/O3_cases/b.e21.BWSSP245cmip6.f09_g17.CMIP6-SSP2-4.5-WACCM.2060-2069.hourlyO3.002, please help me in modifying user_nl_datm_streams file.

**Comment by olyson** [comment]
Did you look at the user_nl_datm_streams in the case I sent you?

/glade/work/oleson/ctsm5.1.dev112/cime/scripts/ctsm51_ctsm51d112_1deg_GSWP3V1_hourlyO3_SSP245/user_nl_datm_streams

**Reply by jyoti1singh** [reply]
Sorry I missed the two uncommented lines. I added the two lines from your case directory and the model build successfully with the hourly ozone inputs. Thanks.

**Comment by jyoti1singh** [comment]
I ran the case (/glade/work/jyotis/O3_CLM5/O3_cases/b.e21.BWSSP245cmip6.f09_g17.CMIP6-SSP2-4.5-WACCM.2060-2069.hourlyO3.002) and it stopped with an error (below is the tail of the cesm log file)
`
.
.
4:MPT ERROR: Rank 4(g:4) is aborting with error code -1.
4:      Process ID: 62833, Host: r4i2n4, Program: /glade/scratch/jyotis/b.e21.BWSSP245cmip6.f09_g17.CMIP6-SSP2-4.5-WACCM.2060-2069.hourlyO3.002/bld/cesm.exe
4:      MPT Version: HPE MPT 2.22  03/31/20 15:59:10
4:
4:MPT: --------stack traceback-------
-1:MPT ERROR: MPI_COMM_WORLD rank 35 has terminated without calling MPI_Finalize()
-1:     aborting job
                      `
I do not understand this error, would you please look into the error.

Thanks,
Jyoti

**Comment by olyson** [comment]
Further up in the cesm log I see this:

0:Abort with message NetCDF: Attempt to use feature that was not turned on when netCDF was built. in file /glade/scratch/vanderwb/hpci-stack/220310-1705-1/54409/pio-2.5.6/src/clib/pioc_support.c at line 2830

I think that this likely means that you need to convert your ozone file from "netCDF-4 classic model" to "cdf5".
E.g.,

nccopy -k cdf5 b.e21.BWSSP245cmip6.f09_g17.CMIP6-SSP2-4.5-WACCM.006.cam.h4.O3_SRF.2055010100-2065010100.nc b.e21.BWSSP245cmip6.f09_g17.CMIP6-SSP2-4.5-WACCM.006.cam.h4.O3_SRF.2055010100-2065010100.cdf5.nc

Or there is another file you are using that is netCDF-4 classic model
